official/projects/roformer/roformer_attention.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Roformer attention layer"""
+# pylint: disable=g-classes-have-attributes
+import math
+
+import tensorflow as tf
+
+EinsumDense = tf.keras.layers.experimental.EinsumDense
+MultiHeadAttention = tf.keras.layers.MultiHeadAttention
+
+
+def _build_trig_vector(length, key_dim):
+    tf_dtype = tf.keras.mixed_precision.global_policy().compute_dtype
+    position_ids = tf.cast(tf.range(length), dtype=tf_dtype)
+    position_ids = tf.expand_dims(position_ids, axis=0)
+    steps = key_dim // 2
+    indices = tf.cast(tf.range(steps), dtype=tf_dtype)
+    indices = tf.pow(tf.constant(10000.0, dtype=tf_dtype), -2 * indices / steps)
+    vec = tf.einsum('bl,d->bld', position_ids, indices)
+    sin_vec = tf.repeat(tf.sin(vec), repeats=2, axis=-1)
+    cos_vec = tf.repeat(tf.cos(vec), repeats=2, axis=-1)
+    sin_vec, cos_vec = tf.expand_dims(sin_vec, 2), tf.expand_dims(cos_vec, 2)
+    return sin_vec, cos_vec
+
+
+@tf.keras.utils.register_keras_serializable(package="Text")
+class RoformerAttention(tf.keras.layers.MultiHeadAttention):
+  def __init__(self,
+               q_max_sequence_length,
+               kv_max_sequence_length,
+               output_range=None,
+               **kwargs):
+    """
+    Args:
+      q_max_sequence_length: maximum length in input for the query
+      kv_max_sequence_length: maximum length in input for key and value, can be different from q_max_sequence_length
+      output_range: length of the query tensor to consider.
+    """
+    super().__init__(**kwargs)
+    self._q_max_sequence_length = q_max_sequence_length
+    self._kv_max_sequence_length = kv_max_sequence_length
+    assert self._key_dim % 2 == 0
+    q_sin_vec, q_cos_vec = _build_trig_vector(self._q_max_sequence_length, self._key_dim)
+    k_sin_vec, k_cos_vec = _build_trig_vector(self._kv_max_sequence_length, self._key_dim)
+    self.q_sin_vec, self.q_cos_vec = (q_sin_vec, q_cos_vec) if output_range is None else (
+        q_sin_vec[:, 0:output_range, ...], q_cos_vec[:, 0:output_range, ...])
+    self.k_sin_vec, self.k_cos_vec = (k_sin_vec, k_cos_vec)
+
+  def roformer_recompute_qkv(self,
+                             q,
+                             k,
+                             v):
+    q_shape = tf.shape(q)
+    q_len = q_shape[1]
+    k_shape = tf.shape(k)
+    k_len = k_shape[1]
+
+    q2 = tf.stack([-q[..., 1::2], q[..., ::2]], axis=4)
```

Comment:
In my understanding, this should not be a fast implementation, as we do the slicing then contact. Maybe we can do a [1, -1] repeat and then masking?